### DISCUSSION TITLE
Can uponSignal take a variadic lambda?

### DISCUSSION THREAD

**Discussion by Radagan** [Kistler-Group/sdbus-cpp#195] [Q&A] [answered]
I have a case where I don't control the signal I'm receiving and it varies in what is passed. I'm trying to do something like what is below, but getting compiler issues. Can you recommend how to handle this case? 

What I have so far...

```
template<typename... arg_types> void cleanup_signal(arg_types... args) {
    std::vector<std::any> arg_list = {args ...};
    for (auto& arg : arg_list)
    {
        if(arg.type() == typeid(const std::string&)) {
            on_signal(std::any_cast<const std::string&>(arg));
            return;
        }
    }
}
```

Used by:

```
proxy = sdbus::createProxy(destination_name, object_path);
proxy->uponSignal(signal).onInterface(interface_name).call([this](auto... args) {
    cleanup_signal(args...);
});
```

**Comment by Radagan** [comment]
As an alternative, is there a way to catch an exception when this passed lambda is called with incorrect arguments?

**Comment by sangelovic** [comment, marked as answer]
@Radagan This isn't possible. The callback must have a concrete signature, because this signature is the source of information for sdbus-c++ introspection mechanisms to find out what data and of what type to read from the incoming signal. sdbus-c++ can't do the message reading in a completely generic way.

I have not yet seen the case of a D-Bus message with data of varying type. I'm not sure if this is correct design. Typically, such messages use the `Variant` type (the type which is there for exactly such cases). Then the client introspects the variant data to see what's inside for that specific D-Bus message.

However, if you don't control the contents of D-Bus signal, the only option for you is to use basic sdbus-c++ API. Like in the client-side example here: https://github.com/Kistler-Group/sdbus-cpp/blob/master/docs/using-sdbus-c%2B%2B.md#implementing-the-concatenator-example-using-basic-sdbus-c-api-layer

**Reply by Radagan** [reply]
@sangelovic Thank you. In my case, I don't control it, so thank you also for the link!

**Comment by sangelovic** [comment]
@Radagan Yes, `sdbus::Error` exception, containing message "Failed to deserialize a XYZ value", is thrown when sdbus-c++ tries to read data of XYZ type from a D-Bus message while there is no data or data of another type. And you can catch that exception if you're running the D-Bus event loop yourself (in your thread), for example like this:

```c++
while(true)
{
    try
    {
       connection->enterEventLoop(); // Blocking
    }
    catch (const sdbus::Error& e)
    {
        processException(e); // Process sdbus-c++ error and then enter the event loop again
    }
}
```

**Reply by sangelovic** [reply]
@Radagan I've converted the issue in a discussion. Please use discussions in the future if you have questions to ask or ideas to share...

**Reply by Radagan** [reply]
> @Radagan I've converted the issue in a discussion. Please use discussions in the future if you have questions to ask or ideas to share...

Sorry about that--will do.
